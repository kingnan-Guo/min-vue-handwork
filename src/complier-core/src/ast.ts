// 枚举 一般使用大写的昂是
export const enum NodeTypes {
    INTERPOLATION,
    SIMPLE_EXPRESSION,
    ELEMENT,
    TEXT,
    ROOT
}