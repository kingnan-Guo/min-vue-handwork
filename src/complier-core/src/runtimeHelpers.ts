export const TO_DISPLAY_STRING = Symbol("toDisplayString")
export const CREATE_ELEMENT_VNODE = Symbol("createElementVNode")

// TO_DISPLAY_STRING 转换成string
export const helperMapName = {
    [TO_DISPLAY_STRING]:"toDisplayString",
    [CREATE_ELEMENT_VNODE]:"createElementVNode",
}
