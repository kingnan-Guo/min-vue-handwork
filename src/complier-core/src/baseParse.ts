import { NodeTypes } from "./ast";
export function baseParse(content:string) {

    const context = createParserContext(content)
    return createRoot(parserChildren(context))
    
    
    // return {
    //     children:[{
    //         type: "interpolation",
    //         content: {
    //             type: "simple_expression",
    //             content:"message"
    //         }
    //     }]
    // }
}




/**
 * 创建 跟节点
 * @param children 
 * @returns 
 */
function createRoot(children) {
    return{
        children
    }
}


function parserChildren(context) {
    const nodes:any = []
    let node
    if (context.source.startsWith("{{")) {
        node = parserInterpolation(context)
        nodes.push(node)
    }
    return nodes

}



/**
 * 解析插值
 */
function parserInterpolation(context) {
    // {{message}}

    const openDelimiter = "{{"
    const closeDelimiter = "}}"
    // 从第二个值开始查找 
    const closeIndex = context.source.indexOf(closeDelimiter, openDelimiter.length)
    // 推进
    // context.source = context.source.slice(openDelimiter.length)
    advanceBy(context, openDelimiter.length)

    const rawContentLenth = closeIndex - openDelimiter.length;
    // content 是去除双括号的值
    const rawContent = context.source.slice(0, rawContentLenth)
    // 去除空格
    const content = rawContent.trim()

    // context.source = context.source.slice(rawContentLenth + closeDelimiter.length)
    advanceBy(context, rawContentLenth + closeDelimiter.length )

    // console.log("context =", context);

    return {
        type: NodeTypes.INTERPOLATION,// "interpolation",
        content: {
            type: NodeTypes.INTERPOLATION,
            content:content
        }
    }
}


function advanceBy(context, length) {
    // context.source.indexOf(context, length)
    context.source = context.source.slice(length)

}

/**
 * 
 * @param content 
 * @returns 
 */
function createParserContext(content) {
    return {
        source:content
    }
}