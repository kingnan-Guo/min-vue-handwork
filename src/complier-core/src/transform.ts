import { NodeTypes } from "./ast";

/**
 * 
 * @param root 
 *  1、遍历 深度优先搜索 traverseNode
 *  2、修改 text的 content
 */
export function transform(root, options:any = {}) {
    const context = creatTransformContext(root, options);
    traverseNode(root, context);
    createRootCodegen(root)

}



function creatTransformContext(root: any, options: any) {
    const context = {
        root,
        nodeTransforms:options.nodeTransforms || []
    }
    return context;
}


function createRootCodegen(root: any) {
    root.codegenNode = root.children[0];
}

function traverseNode(node: any, context: any) {

    const nodeTransforms = context.nodeTransforms
    for (let j = 0; j < nodeTransforms.length; j++) {
        const transform = nodeTransforms[j];
        transform(node);
        
    }
    // if (node.type === NodeTypes.TEXT) {
    //     node.content = node.content + "hellow"
    // }

    traverseChildren(node,  context)
}

function traverseChildren(node: any, context: any) {
    const children = node.children
    if (children) {
        for (let i = 0; i < children.length; i++) {
            const node = children[i];

            traverseNode(node, context)
            
        }
    }
}


