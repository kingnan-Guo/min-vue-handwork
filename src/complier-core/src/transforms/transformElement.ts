// 把处理element类型

import { NodeTypes, createVnodeCall } from "../ast";
// import { CREATE_ELEMENT_VNODE } from "../runtimeHelpers";


   
export function transformElement(node, context) {
    if (node.type === NodeTypes.ELEMENT) {
        return () => {

            // context.helper(CREATE_ELEMENT_VNODE) //将 helps 整理到 ast/createVnodeCall 中


            /**
             * 中间处理层
             * 1、处理tag
             * 2、处理props 
             */
            // tag
            const vnodeTag = `'${node.tag}'`
            let vnodeProps;
            const children = node.children
            let vnodeChildren = children[0]
            // const vnodeElement = {
            //     type: NodeTypes.ELEMENT,
            //     tag: vnodeTag,
            //     props: vnodeProps,
            //     children: vnodeChildren
            // }
    
            /**
             * 因为在最初 处理的时候 处理的额 是 node。codegenNode
             * 为了 可以处理 到 vnodeElement ，所以赋值到 node.codegenNode
             * 将codegenNode 体换的目的 是逐级处理
             */
            node.codegenNode = createVnodeCall(context, vnodeTag, vnodeProps, vnodeChildren) // return vnodeElement


        }
    }
}
