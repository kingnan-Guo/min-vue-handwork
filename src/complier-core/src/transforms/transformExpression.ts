import { NodeTypes } from "../ast";

/**
 * 
 * @param node 
 */
export function transformExpression(node) {
    if (node.type === NodeTypes.INTERPOLATION) {
        
        const rawContent = processExperssion(node.content);
        node.content = rawContent
    }
}

function processExperssion(node: any) {
    node.content = `_ctx.${node.content}`
    return node
}
