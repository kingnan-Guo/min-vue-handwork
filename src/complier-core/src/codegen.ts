import { NodeTypes } from "./ast";
import { TO_DISPLAY_STRING, helperMapName } from "./runtimeHelpers";
/**
 * 
 * @param root 
 *  1、遍历 深度优先搜索 traverseNode
 *  2、修改 text的 content
 */
export function generate(ast: any) {
    const context = createCodegenContext();
    const { push } = context;

    // let code = "";
    // code += "return "
    const functionName = "render"
    const args = ["_ctx", "_cache"]
    const singnature = args.join(", ");

    /**
     * genFunctionPreamble
     * 
     * const VueBinging = "Vue";
     * // helpers 基于类型的处理 添加 不同的额 函数，当处理插值的时候 使用 helper
     * // 期望在ast 上直接就带有  helpers 属性
     * const helpers =["toDisplayString"];
     * const aliasHelp = (s) => `${s}: _${s}`
     * // push(`const { ${helpers.map(e => aliasHelp(e)).join(", ")} } = ${VueBinging} `);
     * push(`const { ${helpers.map(aliasHelp).join(", ")} } = ${VueBinging}`);
     * push("\n");
     */

    genFunctionPreamble(ast, context);

    
    // const node = ast.codegenNode
    // code += `function ${functionName}(${singnature}){`
    // code += `renturn `
    // code += genNode(ast, code)
    // code += `}`

    
    push(`function ${functionName}(${singnature}){`)
    push(`return `)
    genNode(ast.codegenNode, context)
    push("}")


    

    return {
        code: context.code,
    }
}

function createCodegenContext() {
    const context = {
        code: "",
        push(source){
            context.code +=source
        },
        helper(key){
            return `_${helperMapName[key]}`
        }
    }
    return context;
}




function genNode(node: any, context: any) {
    // 分类处理
    switch (node.type) {
        case NodeTypes.INTERPOLATION:
            geninterpolation(node, context)
            break;
        case NodeTypes.TEXT:
            // const { push } = context;
            // push(` '${node.content}'`);
            genText(node, context)
            break;
        case NodeTypes.SIMPLE_EXPRESSION:
            genExpression(node, context)
            break;


        default:
            break;
    }

}




/**
 * genFunctionPreamble
 * @param ast 
 * @param context 
 * 
 * 
 * const VueBinging = "Vue";
 * // helpers 基于类型的处理 添加 不同的额 函数，当处理插值的时候 使用 helper
 * // 期望在ast 上直接就带有  helpers 属性
 * const helpers =["toDisplayString"];
 * const aliasHelp = (s) => `${s}: _${s}`
 * // push(`const { ${helpers.map(e => aliasHelp(e)).join(", ")} } = ${VueBinging} `);
 * push(`const { ${helpers.map(aliasHelp).join(", ")} } = ${VueBinging}`);
 * push("\n");
 * 
 */
function genFunctionPreamble(ast: any, context: { code: string; push(source: any): void; }) {
    // throw new Error("Function not implemented.");
    const { push } = context;
    const VueBinging = "Vue";
    // helpers 基于类型的处理 添加 不同的额 函数，当处理插值的时候 使用 helper
    // 期望在ast 上直接就带有  helpers 属性
    const helpers =[TO_DISPLAY_STRING];
    const aliasHelp = (s) => `${helperMapName[s]}: _${helperMapName[s]}`;
    if (ast.helpers.length > 0) {
        push(`const { ${helpers.map(aliasHelp).join(", ")} } = ${VueBinging}`);
    }
    push("\n");
    push(`return `);


}

function geninterpolation(node: any, context: any) {
    const { push, helper } = context;
    // push(`_toDisplayString(_ctx.message)`);
    // push(`${helperMapName[TO_DISPLAY_STRING]}(`);
    push(`${helper(TO_DISPLAY_STRING)}(`);
    console.log("node ===", node);
    
    genNode(node.content, context)
    push(`)`)

}

function genText(node: any, context: any) {
    const { push } = context;
    push(` '${node.content}'`);
}

function genExpression(node: any, context: any) {
    const { push } = context;
    push(`${node.content}`);
}

