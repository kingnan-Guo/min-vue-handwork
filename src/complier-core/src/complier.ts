/**
 * complier-core 统一出口
 */

import { baseParse } from "./baseParse";
import { generate } from "./codegen";
import { transform } from "./transform";
import { transformElement } from "./transforms/transformElement";
import { transformExpression } from "./transforms/transformExpression";
import { transformText } from "./transforms/transformText";

export function baseCompile(template) {
    const ast = baseParse(template)
        
    transform(ast, {
        // 有顺序  先将 更改节点 然后再 更改 text
        nodeTransforms: [transformExpression,transformElement, transformText]
    });
    console.log("ast ==",ast);
    return generate(ast)
}