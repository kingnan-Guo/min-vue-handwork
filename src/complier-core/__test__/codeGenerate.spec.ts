import { baseParse } from "../src/baseParse";
import { NodeTypes } from "../src/ast";
import { transform } from "../src/transform";
import { generate } from "../src/codegen";
import { transformExpression } from "../src/transforms/transformExpression";



describe("code Generate ", () => {

    
    it("string ", () => {
        // const ast = baseParse("<div>king-{{message}}</div>")
        const ast = baseParse("hi1")
        
        transform(ast);
        console.log("ast ==",ast);
        const { code } = generate(ast)
        // 外部扩展处理 NodeTypes.TEXT

        // 快照测试 ：给code 拍照后续对比 出bug
        expect(code).toMatchSnapshot()

        


    })

    it("interpolation", () => {
        const ast = baseParse("{{message}}")
        
        transform(ast, {
            nodeTransforms: [transformExpression]
        });
        console.log("ast ==",ast);
        const { code } = generate(ast)
        // 快照测试 ：给code 拍照后续对比 出bug
        expect(code).toMatchSnapshot()
    })


});