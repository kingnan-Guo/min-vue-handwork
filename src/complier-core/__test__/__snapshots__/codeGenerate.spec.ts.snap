// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`code Generate  element 1`] = `
"const { creatElementVNode: _creatElementVNode } = Vue
return function render(_ctx, _cache){return _creatElementVNode('div', null, null)}"
`;

exports[`code Generate  element-all 1`] = `
"const { toDisplayString: _toDisplayString, creatElementVNode: _creatElementVNode } = Vue
return function render(_ctx, _cache){return _creatElementVNode('div', null, 'king-' + _toDisplayString(_ctx.message))}"
`;

exports[`code Generate  element-all-test 1`] = `
"const { creatElementVNode: _creatElementVNode, toDisplayString: _toDisplayString } = Vue
return function render(_ctx, _cache){return _creatElementVNode('div', null,_creatElementVNode(span, null,))}"
`;

exports[`code Generate  interpolation 1`] = `
"const { toDisplayString: _toDisplayString } = Vue
return function render(_ctx, _cache){return _toDisplayString(_ctx.message)}"
`;

exports[`code Generate  string  1`] = `
"
return function render(_ctx, _cache){return  'hi1'}"
`;
